Add SPI NOR configuration generator

kowalewskijan opened the pull request:
These changes contain tools able to generate JSON with SPI NOR configurations for various of chips.
Configurations are dumped from such tools as:
* Linux SPI NOR driver
* Flashrom
* OpenOCD

Tool (**spi_nor_json_gen.py**) downloads, patches, builds and runs modules to generate JSONs, then merges into one final JSON so we can use it to configure litespi nor controller.

@enjoy-digital @piotr-binkowski @mithro please take a look.

mithro commented on CONTRIBUTORS:
```diff
@@ -9,3 +9,4 @@ If you think you should be in this list and don't find yourself, write to floren
 and we'll fix it!
 
 Contributors:
+Copyright (c) 2020 Antmicro <www.antmicro.com>
```
No `Copyright (c) 2020` here. 

Probably should add myself and Florent here too.

kowalewskijan replied:
fixed

mithro commented on tools/spi_nor_config_generator/flashrom_json_gen.patch:
```diff
@@ -0,0 +1,1018 @@
+From bf47bd64c9fd156fa7aa6d5c2ecfadba01fe5b12 Mon Sep 17 00:00:00 2001
```
I don't get why this patch exists?

mithro replied:
There was a script at https://github.com/mithro/litex/blob/spi-module/litex/soc/cores/spi/pull_flashrom_data.py which parsed this file directly.

kowalewskijan replied:
I added pull_flashrom_data.py so the patch is not needed anymore.

mithro commented:
Please take a look at the code at https://github.com/mithro/litex/blob/spi-module/litex/soc/cores/spi/pull_flashrom_data.py and take from there.

kowalewskijan commented:
I added pull_flashrom_data.py and changed it so it can generate JSON.
Additionally I added an option to generate Python classes from parsed configs.

kowalewskijan commented:
I made generating final JSON optional - Python classes generation is now mandatory.
Updated **litespi/modules.py** with generated classes.

mithro commented on CONTRIBUTORS:
```diff
@@ -9,3 +9,4 @@ If you think you should be in this list and don't find yourself, write to floren
 and we'll fix it!
 
 Contributors:
+Antmicro <www.antmicro.com>
```
Please a me (`Tim 'mithro' Ansell <me@mith.ro>`) and Florent to this list.

mithro commented on litespi/modules.py:
```diff
@@ -0,0 +1,12300 @@
+# Generated using 'spi_nor_cfg_gen.py'
+
+class AT25FS010(object):
```
Please make all the spi flash modules inherit from a common base module (which is well documented).

kowalewskijan replied:
I added common base module [here](https://github.com/litex-hub/litespi/pull/1/files?file-filters%5B%5D=.patch&file-filters%5B%5D=.py&file-filters%5B%5D=No+extension&file-filters%5B%5D=dotfile&hide-deleted-files=true#diff-0114c117482b2e8cd170c5a98f7a90d1)

mithro commented on litespi/modules.py:
```diff
@@ -0,0 +1,12300 @@
+    total_size =     131072   # bytes
+    page_size  =        256   # bytes
+
+    access_modes = {
```
Please use enum values here.

kowalewskijan replied:
Classes have now **supported_modes** field which is **SpiNorConfig** type defined [here](https://github.com/litex-hub/litespi/pull/1/files?file-filters%5B%5D=.patch&file-filters%5B%5D=.py&file-filters%5B%5D=No+extension&file-filters%5B%5D=dotfile&hide-deleted-files=true#diff-6f4a7670a48198f06e6250d0ccda1601)

kowalewskijan replied:
**SpiNorConfig** replace with **SpiNorFeatures**

mithro commented on litespi/modules.py:
```diff
@@ -0,0 +1,12300 @@
+    addr32_support = False
+
+    dummy_bits = 8
+
```
Two new lines between classes.

mithro commented on litespi/modules.py:
```diff
@@ -0,0 +1,12300 @@
+        8 : False,
+    }
+    fast_read_support = True
+    addr32_support = False
```
Why is addr32_support and fast_read_support seperate from the access_modes and addr_widths?

kowalewskijan replied:
Fixed it. They are now part of [SpiNorConfig](https://github.com/litex-hub/litespi/pull/1/files?file-filters%5B%5D=.patch&file-filters%5B%5D=.py&file-filters%5B%5D=No+extension&file-filters%5B%5D=dotfile&hide-deleted-files=true#diff-6f4a7670a48198f06e6250d0ccda1601)

kowalewskijan replied:
**SpiNorConfig** replaced with **SpiNorFeatures**

mithro commented on litespi/modules.py:
```diff
@@ -0,0 +1,12300 @@
+
+class AT25FS010(object):
+
+    manufacturer_id = 0x1f
```
Please use enum for the manufacturer_id.

mithro commented on tools/spi_nor_config_generator/pull_flashrom_data.py:
```diff
@@ -0,0 +1,461 @@
+regex_list_end = re.compile('}([,\\s]*)}')
+
+
+import enum
```
This enum should be part of the litespi library (rather than included here).

kowalewskijan replied:
I moved it right [here](https://github.com/litex-hub/litespi/pull/1/files?file-filters%5B%5D=.patch&file-filters%5B%5D=.py&file-filters%5B%5D=No+extension&file-filters%5B%5D=dotfile&hide-deleted-files=true#diff-e07142bb7261ca9dcc8fc3ec66b0ecef)

mithro commented on tools/spi_nor_config_generator/spi_nor_cfg_gen.py:
```diff
@@ -0,0 +1,463 @@
+# Some features could be not mentioned in modules we parse from.
+# In case there is a chip which supports a feature but it was not
+# mentioned, we can override config here.
+override_chip_cfg = {
```
Can we store this override in a separate file.

kowalewskijan replied:
I moved it right [here](https://github.com/litex-hub/litespi/pull/1/files?file-filters%5B%5D=.patch&file-filters%5B%5D=.py&file-filters%5B%5D=No+extension&file-filters%5B%5D=dotfile&hide-deleted-files=true#diff-0a6347bd6f10ebc4431f61b033d99b5a) and additionally I changed override configs to use enums

mithro commented on tools/spi_nor_config_generator/spi_nor_cfg_gen.py:
```diff
@@ -0,0 +1,463 @@
+            if entry["chip_name"][0].isdigit():
+                name = 'X' + name
+
+            generated_class = '''class {pn}(object):
```
Can we move this generated class to a seperate template file?

kowalewskijan replied:
Function moved [here](https://github.com/litex-hub/litespi/pull/1/files?file-filters%5B%5D=.patch&file-filters%5B%5D=.py&file-filters%5B%5D=No+extension&file-filters%5B%5D=dotfile&hide-deleted-files=true#diff-0a7a8b224ab31672a326a8755a06eb3f)

enjoy-digital commented:
@mithro: i had a quick look at the PR, but since you already started working on that some time ago, it will probably be better that you merge it when you are happy with it. We just need to be careful the initial structure of the classes will not move (or not move a lot) in the future, because that's possible we'll add some SPI module definitions that are not present in the others projects and we'll also want to be able to update the SPI module definitions when new definitions are added to the others projects, so we need to be able to mix generated definitions and manually added definitions easily. Not sure we'll add that much definitions ourselves (since the others project are probably covering most parts) but that's worth thinking about it now and how we'll handle that if the case happens.

kowalewskijan commented on litespi/modules/__init__.py:
```diff
@@ -0,0 +1,2 @@
+from litespi.modules.generated_modules import *
+from litespi.modules.modules import *
```
I added _modules_ directory where we can store auto-generated classes and add modules manually in the other file.

mithro replied:
That sounds like a good solution!

mithro commented on litespi/spi_nor_flash_module.py:
```diff
@@ -0,0 +1,131 @@
+    bus_width : int
+        expected number of data lines
+    """
+    def _check_xfer_mode(self, cfg):
```
This changes `self` values, hence it should be called something like `_set_xfer_mode` or `_workout_best_xfer_mode` etc.

This function should have a docstring.

kowalewskijan replied:
I renamed it to [_configure_transfer_mode](https://github.com/litex-hub/litespi/pull/1/files#diff-0114c117482b2e8cd170c5a98f7a90d1R58) and I added a docstring.

mithro commented on litespi/spi_nor_flash_module.py:
```diff
@@ -0,0 +1,131 @@
+
+    bus_width : int
+        expected number of data lines
+    """
```
Newline between class docstring and the `def`.

mithro commented on litespi/spi_nor_flash_module.py:
```diff
@@ -0,0 +1,131 @@
+
+    def __init__(self, spi_nor_cfg=SpiNorConfig.SINGLE_MODE):
+        # Check if mandatory attributes are set by an inheritor
+        assert 'manufacturer_id' in dir(self)
```
```python
assert hasattr(self, 'device_id')
```

mithro commented on litespi/spi_nor_config.py:
```diff
@@ -0,0 +1,29 @@
+    """SPI NOR Config
+
+    The ``SpiNorConfig`` class provides enums used to describe SPI NOR
+    configuration and chip capabilites.
```
Isn't this covered by the `SpiNorFeatures` class?

kowalewskijan replied:
Partially yes, there are some missing features. I will add them there and swap SpiNorConfig to SpiNorFeatures

mithro replied:
I think I had most of the important features at https://github.com/mithro/litex/blob/d7184b97494012a8c21ee93df231b758a61e9dec/litex/soc/cores/spi/modules.py#L141-L193 ?

kowalewskijan replied:
That's correct, but I would like to add to `SpiNorFeatures` class such flags as:
- Dual mode support
- Octal mode support
- Fast read mode (24bit)

In `SpiNorFeatures` there is a flag for Quad Mode (`FEATURE_QPI`) so adding `FEATURE_DUAL` and `FEATURE_OCTAL` would be complementary.
Additionally there is `FEATURE_4BA_FAST_READ` but it lacks of 24-bit fast read.

mithro replied:
Dual mode is not clear. There are multiple "dual" modes;
 -     `READ_1_1_2        = _Op(0x3b, "Read data bytes (Dual Output SPI)")`
 -     `READ_1_2_2        = _Op(0xbb, "Read data bytes (Dual I/O SPI)")`
 -     `READ_1_2_2_DTR    = _Op(0xbd, "")`

kowalewskijan replied:
Will be reworked in the next PR

mithro commented on litespi/spi_nor_flash_module.py:
```diff
@@ -0,0 +1,131 @@
+            cmd_picker = ("SpiNorFlashOpCodes.READ%s%s" %
+                          (fast, addr32))
+        else:
+            cmd_picker = ("SpiNorFlashOpCodes.READ_%d_%d_%d%s" %
```
`getattr`

kowalewskijan commented:
I added missing [MetaSizes](https://github.com/litex-hub/litespi/pull/1/files#diff-0114c117482b2e8cd170c5a98f7a90d1R5) class and I added [print_modules](https://github.com/litex-hub/litespi/pull/1/files#diff-df41fe3aa0d9e2aeb6bb909f44e97af9R5) function.

mithro commented on litespi/modules/generated_modules.py:
```diff
@@ -0,0 +1,8197 @@
+    page_size   =        256   # bytes
+    total_pages =        512
+
+    supported_modes = SpiNorFeatures.FEATURE_ADDR_WIDTH_1 | \
```
Better wrapping would be;
```python
    supported_modes = \
		SpiNorFeatures.FEATURE_ADDR_WIDTH_1 | \
		SpiNorFeatures.FEATURE_FAST_READ | \
		SpiNorFeatures.FEATURE_SINGLE 
```

mithro commented on litespi/spi_nor_flash_module.py:
```diff
@@ -0,0 +1,286 @@
+from litespi.opcodes import SpiNorFlashOpCodes
+
+
+class MetaSizes(type):
```
Probably worth putting this in it's own file.

Remove the debugging prints.

kowalewskijan replied:
I removed debug prints.

mithro commented on litespi/spi_nor_features.py:
```diff
@@ -0,0 +1,75 @@
+
+    FEATURE_WRSR_EITHER     = FEATURE_WRSR_EWSR | FEATURE_WRSR_WREN
+
+    FEATURE_SINGLE          = enum.auto()
```
Can we just have a "supported opcodes" type thing or "supported read opcodes"?

kowalewskijan replied:
Will be done in the next PR

mithro commented:
I think this is looking pretty good. If you want to do something around the "supported opcodes" moving metasizes into it's own file then I think it is fine to land and continue to iterate in a new pull request.

kowalewskijan commented:
I squashed the commits to have a clear history and fixed some minor issues.
I'll prepare separate PR with possible changes around "supported opcodes" and moving metasizes.

kowalewskijan commented:
@mithro can we merge it? I'll add changes around MetaSizes and supported_opcodes in the next PR.

mithro commented:
SGTM!

mithro commented on tools/spi_nor_config_generator/spi_nor_cfg_gen.py:
```diff
@@ -0,0 +1,380 @@
+
+
+def main():
+    parser = argparse.ArgumentParser(description='\
```
You should be using a triple quoted string for this.
